File: src/structure/resolveConflicts.ts
import path from "node:path";
import {Configuration, ConflictResolutionStrategy} from "../configuration/configuration.js";
import {DerivedFile, GeneratedFile} from "../converter/generated.js";
import {normalizeItems} from "../utils/normalizeItems.js";
import {toPosix} from "../utils/path.js";
import {StructureItem} from "./structure.js";
import {packageToOutputFileName} from "./package/packageToFileName.js";
import {createTargetFile} from "./createTargetFile.js";
import {createGeneratedFile} from "./createGeneratedFile.js";

export interface TargetFile extends StructureItem {
    body: string
}

function resolveConflictResolutionStrategy(
    outputFileName: string,
    configuration: Configuration,
): ConflictResolutionStrategy | undefined {
    const {conflictResolutionStrategy} = configuration

    for (const [pattern, strategy] of Object.entries(conflictResolutionStrategy)) {
        const regexp = new RegExp(pattern)

        if (regexp.test(outputFileName)) {
            return strategy;
        }
    }

    return undefined
}

function handleConflicts<T>(
    conflicts: Record<string, T[]>,
    type: string,
    printItemInfo: (item: T) => string,
) {
    if (Object.keys(conflicts).length) {
        for (const [outputFileName, items] of Object.entries(conflicts)) {
            console.error(`Conflicting ${type} File: ${toPosix(outputFileName)}`);

            for (const index in items) {
                const item = items[index]

                console.error(`--- Item #${Number(index) + 1} ---`);
                console.error(printItemInfo(item))
            }
        }

        throw new Error("There are conflicts in file names")
    }
}

function resolveTargetFileConflicts(targetFiles: TargetFile[], configuration: Configuration): TargetFile[] {
    const normalizedTargetFiles = normalizeItems(
        targetFiles,
        (item) => packageToOutputFileName(
            item.package,
            item.fileName,
            configuration
        ),
        (outputFileName, item, other) => {
            const conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration) ?? "join"

            if (conflictResolutionStrategy === "error") {
                return null
            }

            if (conflictResolutionStrategy === "replace") {
                return other
            }

            if (conflictResolutionStrategy === "join") {
                if (
                    item.moduleName !== other.moduleName
                    || item.qualifier !== other.qualifier
                ) {
                    // items are incompatible
                    return null
                }

                return {
                    ...item,
                    hasRuntime: item.hasRuntime || other.hasRuntime,
                    body: `${item.body}\n\n${other.body}`,
                }
            }

            throw new Error(`Unknown namespace strategy: ${conflictResolutionStrategy}`)
        }
    )

    handleConflicts(normalizedTargetFiles.conflicts, "Target", item => {
        const module = `Module: ${item.moduleName}`
        const qualifier = item.qualifier ? `\nQualifier: ${item.qualifier}` : ""
        return `${module}${qualifier}\nContent:\n${item.body}`
    })

    return normalizedTargetFiles.items
}

function resolvePrimaryFileConflicts(primaryFiles: GeneratedFile[], configuration: Configuration): GeneratedFile[] {
    const normalizedPrimaryFiles = normalizeItems(
        primaryFiles,
        (item) => item.fileName,
        (outputFileName, _, other) => {
            const conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration) ?? "error"

            if (conflictResolutionStrategy === "error") {
                return null
            }

            if (conflictResolutionStrategy === "replace") {
                return other
            }

            if (conflictResolutionStrategy === "join") {
                // primary files can not be merged
                return null
            }

            throw new Error(`Unknown namespace strategy: ${conflictResolutionStrategy}`)
        }
    )

    handleConflicts(normalizedPrimaryFiles.conflicts, "Primary", item => {
        return `Content:\n${item.body}`
    })

    return normalizedPrimaryFiles.items
}

function resolveDerivedFilesConflicts(primaryFiles: DerivedFile[], configuration: Configuration): DerivedFile[] {
    const normalizedDerivedFiles = normalizeItems(
        primaryFiles,
        (item) => packageToOutputFileName(
            item.package,
            item.fileName,
            configuration
        ),
        (outputFileName, item, other) => {
            const conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration) ?? "join"

            if (conflictResolutionStrategy === "error") {
                return null
            }

            if (conflictResolutionStrategy === "replace") {
                return other
            }

            if (conflictResolutionStrategy === "join") {
                return {
                    ...item,
                    body: `${item.body}\n\n${other.body}`,
                }
            }

            throw new Error(`Unknown namespace strategy: ${conflictResolutionStrategy}`)
        }
    )

    handleConflicts(normalizedDerivedFiles.conflicts, "Derived", item => {
        return `Content:\n${item.body}`
    })

    return normalizedDerivedFiles.items
}

function resolveCompoundFilesConflicts(
    compoundFiles: GeneratedFile[],
    configuration: Configuration,
): GeneratedFile[] {
    const normalizedCompoundFiles = normalizeItems(
        compoundFiles,
        (item) => item.fileName,
        (outputFileName, item, other) => {
            const conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration) ?? "join"

            if (conflictResolutionStrategy === "error") {
                return null
            }

            if (conflictResolutionStrategy === "replace") {
                // auxiliary file can't replace primary file
                return {
                    ...item,
                    body: `${item.body}${other.body}`,
                }
            }

            if (conflictResolutionStrategy === "join") {
                return {
                    ...item,
                    body: `${item.body}${other.body}`,
                }
            }

            throw new Error(`Unknown namespace strategy: ${conflictResolutionStrategy}`)
        }
    )

    handleConflicts(normalizedCompoundFiles.conflicts, "Compound", item => {
        return `Content:\n${item.body}`
    })

    return normalizedCompoundFiles.items
}

export function resolveConflicts(
    targetFiles: TargetFile[],
    derivedFiles: DerivedFile[],
    generatedFiles: GeneratedFile[],
    configuration: Configuration,
): GeneratedFile[] {
    const {output} = configuration

    const normalizedTargetFiles = resolveTargetFileConflicts(targetFiles, configuration)

    const primaryFiles: GeneratedFile[] = [
        ...normalizedTargetFiles.map(item => {
            const outputFileName = packageToOutputFileName(
                item.package,
                item.fileName,
                configuration
            )

            const fileName = path.resolve(output, outputFileName)
            const body = createTargetFile(item, item.body, configuration)

            return {
                fileName,
                body,
            }
        }),
        ...generatedFiles,
    ]

    const normalizedPrimaryFiles = resolvePrimaryFileConflicts(primaryFiles, configuration)

    const primaryFileNames = new Set(normalizedPrimaryFiles.map(item => item.fileName))

    const normalizedDerivedFiles = resolveDerivedFilesConflicts(derivedFiles, configuration)

    const standaloneDerivedFiles: GeneratedFile[] = []
    const auxiliaryDerivedFiles: GeneratedFile[] = []

    for (const item of normalizedDerivedFiles) {
        const outputFileName = packageToOutputFileName(
            item.package,
            item.fileName,
            configuration
        )

        const fileName = path.resolve(output, outputFileName)

        if (primaryFileNames.has(fileName)) {
            auxiliaryDerivedFiles.push({
                fileName,
                body: item.body
            })
        } else {
            const body = createGeneratedFile(
                item.package,
                item.fileName,
                item.imports,
                item.body,
                configuration
            )

            standaloneDerivedFiles.push({
                fileName,
                body,
            })
        }
    }

    const compoundFiles = [
        ...primaryFiles,
        ...auxiliaryDerivedFiles,
    ]

    const normalizedCompoundFiles = resolveCompoundFilesConflicts(compoundFiles, configuration)

    for (const item of normalizedCompoundFiles) {
        console.log(`Target file: ${toPosix(item.fileName)}`)
    }

    for (const item of standaloneDerivedFiles) {
        console.log(`Generated file: ${toPosix(item.fileName)}`)
    }

    return [
        ...normalizedCompoundFiles,
        ...standaloneDerivedFiles,
    ]
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/resolveConflicts.kt
package io.github.sgrishchenko.karakum.structure

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.configuration.ConflictResolutionStrategy
import io.github.sgrishchenko.karakum.extension.DerivedFile
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.structure.`package`.packageToOutputFileName
import io.github.sgrishchenko.karakum.util.normalizeItems
import io.github.sgrishchenko.karakum.util.toPosix
import js.array.ReadonlyArray
import js.objects.Object
import js.objects.ReadonlyRecord
import kotlinx.js.JsPlainObject
import node.path.path

@JsPlainObject
external interface TargetFile : StructureItem {
    val body: String
}

private fun resolveConflictResolutionStrategy(
    outputFileName: String,
    configuration: Configuration,
): ConflictResolutionStrategy? {
    val conflictResolutionStrategy = configuration.conflictResolutionStrategy

    for ((pattern, strategy) in Object.entries(conflictResolutionStrategy)) {
        val regexp = pattern.toRegex()

        if (regexp.containsMatchIn(outputFileName)) {
            return strategy
        }
    }

    return null
}

private fun <T> handleConflicts(
    conflicts: ReadonlyRecord<String, ReadonlyArray<T>>,
    type: String,
    printItemInfo: (item: T) -> String,
) {
    if (Object.keys(conflicts).isNotEmpty()) {
        for ((outputFileName, items) in Object.entries(conflicts)) {
            console.error("Conflicting $type File: ${toPosix(outputFileName)}")

            for ((index, item) in items.withIndex()) {
                console.error("--- Item #${index + 1} ---")
                console.error(printItemInfo(item))
            }
        }

        error("There are conflicts in file names")
    }
}

private fun resolveTargetFileConflicts(
    targetFiles: ReadonlyArray<TargetFile>,
    configuration: Configuration,
): ReadonlyArray<TargetFile> {
    val normalizedTargetFiles = normalizeItems(
        targetFiles,
        { item ->
            packageToOutputFileName(
                item.`package`,
                item.fileName,
                configuration
            )
        },
        merge@{ outputFileName, item, other ->
            if (item.body == "") {
                return@merge other
            }

            if (other.body == "") {
                return@merge item
            }

            val conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration)
                ?: ConflictResolutionStrategy.join

            if (conflictResolutionStrategy == ConflictResolutionStrategy.error) {
                return@merge null
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.replace) {
                // target files cannot be replaced
                return@merge null
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.join) {
                if (
                    item.moduleName !== other.moduleName
                    || item.qualifier !== other.qualifier
                ) {
                    // items are incompatible
                    return@merge null
                }

                return@merge TargetFile(
                    fileName = item.fileName,
                    `package` = item.`package`,
                    moduleName = item.moduleName,
                    qualifier = item.qualifier,
                    hasRuntime = item.hasRuntime || other.hasRuntime,
                    imports = item.imports,

                    body = "${item.body}\n\n${other.body}",
                )
            }

            error("Unknown conflict resolution strategy: $conflictResolutionStrategy")
        }
    )

    handleConflicts(normalizedTargetFiles.conflicts, "Target") { item ->
        val module = "Module: ${item.moduleName}"
        val qualifier = if (item.qualifier != null) "\nQualifier: ${item.qualifier}" else ""
        "${module}${qualifier}\nContent:\n${item.body}"
    }

    return normalizedTargetFiles.items
}

private fun resolvePrimaryFileConflicts(
    primaryFiles: ReadonlyArray<GeneratedFile>,
    configuration: Configuration
): ReadonlyArray<GeneratedFile> {
    val normalizedPrimaryFiles = normalizeItems(
        primaryFiles,
        { item -> item.fileName },
        merge@{ outputFileName, item, other ->
            if (item.body == "") {
                return@merge other
            }

            if (other.body == "") {
                return@merge item
            }

            val conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration)
                ?: ConflictResolutionStrategy.error

            if (conflictResolutionStrategy == ConflictResolutionStrategy.error) {
                return@merge null
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.replace) {
                // primary files cannot be replaced
                return@merge null
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.join) {
                // primary files cannot be merged
                return@merge null
            }

            error("Unknown conflict resolution strategy: $conflictResolutionStrategy")
        }
    )

    handleConflicts(normalizedPrimaryFiles.conflicts, "Primary") { item ->
        "Content:\n${item.body}"
    }

    return normalizedPrimaryFiles.items
}

private fun resolveDerivedFilesConflicts(
    primaryFiles: ReadonlyArray<DerivedFile>,
    configuration: Configuration,
): ReadonlyArray<DerivedFile> {
    val normalizedDerivedFiles = normalizeItems(
        primaryFiles,
        { item ->
            packageToOutputFileName(
                item.`package`,
                item.fileName,
                configuration
            )
        },
        merge@{ outputFileName, item, other ->
            if (item.body == "") {
                return@merge other
            }

            if (other.body == "") {
                return@merge item
            }

            val conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration)
                ?: ConflictResolutionStrategy.join

            if (conflictResolutionStrategy == ConflictResolutionStrategy.error) {
                return@merge null
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.replace) {
                if (item.body !== other.body) return@merge null
                return@merge other
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.join) {
                return@merge DerivedFile(
                    fileName = item.fileName,
                    `package` = item.`package`,
                    imports = item.imports,

                    body = "${item.body}\n\n${other.body}",
                )
            }

            error("Unknown conflict resolution strategy: $conflictResolutionStrategy")
        }
    )

    handleConflicts(normalizedDerivedFiles.conflicts, "Derived") { item ->
        "Content:\n${item.body}"
    }

    return normalizedDerivedFiles.items
}

private fun resolveCompoundFilesConflicts(
    compoundFiles: ReadonlyArray<GeneratedFile>,
    configuration: Configuration,
): ReadonlyArray<GeneratedFile> {
    val normalizedCompoundFiles = normalizeItems(
        compoundFiles,
        { item -> item.fileName },
        merge@{ outputFileName, item, other ->
            if (item.body === "") {
                return@merge other
            }

            if (other.body === "") {
                return@merge item
            }

            val conflictResolutionStrategy = resolveConflictResolutionStrategy(outputFileName, configuration)
                ?: ConflictResolutionStrategy.join

            if (conflictResolutionStrategy == ConflictResolutionStrategy.error) {
                return@merge null
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.replace) {
                // auxiliary file can't replace primary file
                return@merge GeneratedFile(
                    fileName = item.fileName,
                    body = "${item.body}${other.body}",
                )
            }

            if (conflictResolutionStrategy == ConflictResolutionStrategy.join) {
                return@merge GeneratedFile(
                    fileName = item.fileName,
                    body = "${item.body}${other.body}",
                )
            }

            error("Unknown conflict resolution strategy: $conflictResolutionStrategy")
        }
    )

    handleConflicts(normalizedCompoundFiles.conflicts, "Compound") { item ->
        "Content:\n${item.body}"
    }

    return normalizedCompoundFiles.items
}

fun resolveConflicts(
    targetFiles: ReadonlyArray<TargetFile>,
    derivedFiles: ReadonlyArray<DerivedFile>,
    generatedFiles: ReadonlyArray<GeneratedFile>,
    configuration: Configuration,
): ReadonlyArray<GeneratedFile> {
    val output = configuration.output

    val normalizedTargetFiles = resolveTargetFileConflicts(targetFiles, configuration)

    val primaryFiles = normalizedTargetFiles
        .map { item ->
            val outputFileName = packageToOutputFileName(
                item.`package`,
                item.fileName,
                configuration
            )

            val fileName = path.resolve(output, outputFileName)
            val body = createTargetFile(item, item.body, configuration)

            GeneratedFile(
                fileName = fileName,
                body = body,
            )
        }
        .plus(generatedFiles)
        .toTypedArray()

    val normalizedPrimaryFiles = resolvePrimaryFileConflicts(primaryFiles, configuration)

    val primaryFileNames = normalizedPrimaryFiles.map { it.fileName }.toSet()

    val normalizedDerivedFiles = resolveDerivedFilesConflicts(derivedFiles, configuration)

    val standaloneDerivedFiles = mutableListOf<GeneratedFile>()
    val auxiliaryDerivedFiles = mutableListOf<GeneratedFile>()

    for (item in normalizedDerivedFiles) {
        val outputFileName = packageToOutputFileName(
            item.`package`,
            item.fileName,
            configuration
        )

        val fileName = path.resolve(output, outputFileName)

        if (fileName in primaryFileNames) {
            auxiliaryDerivedFiles += GeneratedFile(
                fileName = fileName,
                body = item.body,
            )
        } else {
            val body = createGeneratedFile(
                item.`package`,
                item.fileName,
                item.imports,
                item.body,
                configuration
            )

            standaloneDerivedFiles += GeneratedFile(
                fileName = fileName,
                body = body,
            )
        }
    }

    val compoundFiles = primaryFiles + auxiliaryDerivedFiles

    val normalizedCompoundFiles = resolveCompoundFilesConflicts(compoundFiles, configuration)

    for (item in normalizedCompoundFiles) {
        console.log("Target file: ${toPosix(item.fileName)}")
    }

    for (item in standaloneDerivedFiles) {
        console.log("Generated file: ${toPosix(item.fileName)}")
    }

    return normalizedCompoundFiles + standaloneDerivedFiles
}
